Auth: Allow expiration of API keys

papagian opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our [`CONTRIBUTING.md`](https://github.com/grafana/grafana/blob/master/CONTRIBUTING.md) guide.
2. Ensure you have added or ran the appropriate tests for your PR.
3. If it's a new feature or config option it will need a docs update. Docs are under the docs folder in repo root.
4. If the PR is unfinished, mark it as a draft PR.
5. Rebase your PR if it gets out of sync with master
6. Name your PR as `<FeatureArea>: Describe your change`. If it's a fix or feature relevant for changelog describe the user  impact in the title. The PR title is used in changelog for issues marked with `add to changelog` label. 
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
-->


<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->
Fixes #17042

This patch introduces a new configuration value for max api key duration. If it is non empty all new keys need to have expiration duration set and it needs to be less than that value. If this value is not set (defaults to -1) you can set the expiration time but do not have to. Existing keys are not affected by configuration value changes.

A new input for new api key lifespan is added in the UI modal, where one can set 1d for example, meaning key is valid 1 day since update (if it is left empty it will mean key won't expire). Also an expiration date column is added in the api key table. In the table only the not expired keys appear.

However the relative api call  optionally accepts a `includeInvalid` parameter and if it’ s true returns all the api keys even the expired ones.

Finally the middleware rejects requests with expired api keys.

**Special notes for your reviewer**:

The patch introduces a new column and a new index (`expires`) in the api_key database table.

There are modifications both in the backend and the frontend.

papagian committed:
Modify backend to allow expiration of API Keys

papagian committed:
Add middleware test for expired api keys

papagian committed:
Modify frontend to enable expiration of API Keys

papagian committed:
Fix frontend tests

papagian committed:
Fix migration and add index for `expires` field

papagian committed:
Add api key tests for database access

papagian committed:
Substitude time.Now() by a mock for test usage

aocenas commented on pkg/api/apikey.go:
```diff
@@ -16,10 +17,15 @@ func GetAPIKeys(c *m.ReqContext) Response {
 	for i, t := range query.Result {
+		expiration := ""
+		if !t.Expires.IsZero() {
+			expiration = t.Expires.Format(time.UnixDate)
```
We do not have to format it here as Time is formatted correctly when marshalled into json and you can set the `expiration` type to `*time.Time` so it can be null. If we need specific format of the date for human readability we should do it on the frontend (for example using local time there instead of string that can be in different timezone than the user)

papagian replied:
I transform it to string because otherwise if `t.Expires` is the zero date and in the frontend the expiration appears to be the January 1, year 1, 00:00:00.000000000 UTC

aocenas replied:
If the `ApiKeyDTO.Expiration` is `*time.Time` instead of `time.Time` you can leave it `nil`. That way it will be sent to frontend empty.

aocenas commented on pkg/services/sqlstore/apikey_test.go:
```diff
@@ -27,5 +28,115 @@ func TestApiKeyDataAccess(t *testing.T) {
+			cmd := m.AddApiKeyCommand{OrgId: 1, Name: "non-expiring", Key: "asd1", SecondsToLive: 0}
+			err := AddApiKey(&cmd)
+			if err != nil {
+				t.Fatalf("expected to work %v", err)
```
Just fyi we have an `github.com/stretchr/testify/assert` util library as dependency so you can use `assert.Nil(err)`.

aocenas commented on public/app/features/api-keys/ApiKeysPage.tsx:
```diff
@@ -170,6 +222,19 @@ export class ApiKeysPage extends PureComponent<Props, any> {
+                  validationEvents={timeRangeValidationEvents}
+                  value={newApiKey.secondsToLive}
+                  onChange={evt =>
+                    this.onApiKeyStateUpdate(evt, ApiKeyStateProps.SecondsToLive, this.transformToSecondsToLive)
```
There is a an issue with having the transform here:
![Kapture 2019-06-20 at 10 49 08](https://user-images.githubusercontent.com/1014802/59834912-16725c80-9349-11e9-9cc9-c26eb1a927be.gif)

This feels jarring and unexpected to me and works differently than the other place we have this. So I think it would be better to do the transform when saving key.

Also that means `Seconds to live` label and `1d` placeholder looks a bit off. Can we change it to `Time to live` and add tooltip saying you can use different time units there `d/m/s/y...`?

papagian replied:
> Also that means Seconds to live label and 1d placeholder looks a bit off. Can we change it to Time to live and add tooltip saying you can use different time units there d/m/s/y...?

do you mean also to remove the placeholder and add a tooltip instead?

aocenas replied:
I think we can leave the placeholder as `1d` and just change the label and add the tooltip. That seems most consistent and informative to me.

torkelo replied:
We have frontend utility that can translate a time range like 10s 5m 15d to milliseconds and vice versa (think they are still in kbn.ts not sure, on mobile now)

Think this should use a time range like that as we have that everywhere else. It can be stored in seconds to make backend easier.

aocenas replied:
You are right https://github.com/grafana/grafana/blob/4b0ad174ff43ba8b629a1c6b3c75051ad7f7c6c0/public/app/core/utils/kbn.ts#L212 seems like exactly what is needed there. There are some almost similar helpers in grafana/ui but probably this should be migrated there too so it is easier to find (though in separate PR)

aocenas reviewed (commented):
Good work, I have a few smaller comments that I think we should address before merge.

aocenas commented on public/app/features/api-keys/ApiKeysPage.tsx:
```diff
@@ -221,6 +287,7 @@ export class ApiKeysPage extends PureComponent<Props, any> {
                   <tr key={key.id}>
                     <td>{key.name}</td>
                     <td>{key.role}</td>
+                    <td>{key.expiration}</td>
```
Hmm for some reason this comment was not saved last time. But here is the place I think we should add the formatting to current browser local time I think instead of having the formatting server side.

@dprokop is it ok to show times in local time or do we have some other way to deal with time formatting?

papagian committed:
Front-end modifications

* Change input label to `Time to live`
* Change input behavior to comply with the other similar
* Add tooltip

papagian committed:
Modify AddApiKey api call response

Expiration should be *time.Time instead of string

papagian committed:
Present expiration date in the selected timezone

papagian committed:
Use kbn for transforming intervals to seconds

papagian committed:
Use `assert` library for tests

marefr commented on pkg/services/sqlstore/migrations/apikey_mig.go:
```diff
@@ -78,4 +78,12 @@ func addApiKeyMigrations(mg *Migrator) {
 	}))
+
+	mg.AddMigration("Add expires to api_key table", NewAddColumnMigration(apiKeyV2, &Column{
+		Name: "expires", Type: DB_DateTime, Nullable: true,
```
I strongly suggest to use a column type of integer instead and store dates as epochs. Reason we've had a lot of problems historically with differences between sqlite, postgres and mysql and our database ORM getting weird results when using a query to check if something has expired by doing a date/time comparison. 

So therefore it's suggested to use an epoch since it's naturally UTC time and removes a lot of headaches. Some examples below.

**user auth tokens:**
 - Migration https://github.com/grafana/grafana/blob/219d711597fb34e53aba065190311af02c887200/pkg/services/sqlstore/migrations/user_auth_token_mig.go#L18-L21
 - Filter by comparing epochs https://github.com/grafana/grafana/blob/219d711597fb34e53aba065190311af02c887200/pkg/services/auth/auth_token.go#L46-L48
https://github.com/grafana/grafana/blob/219d711597fb34e53aba065190311af02c887200/pkg/services/auth/auth_token.go#L110-L114

**login attempts:**
 - Migration https://github.com/grafana/grafana/blob/219d711597fb34e53aba065190311af02c887200/pkg/services/sqlstore/migrations/login_attempt_mig.go#L30
 - FIlter by comparing epochs https://github.com/grafana/grafana/blob/219d711597fb34e53aba065190311af02c887200/pkg/services/sqlstore/login_attempt.go#L73

marefr commented on pkg/services/sqlstore/migrator/dialect.go:
```diff
@@ -49,6 +49,9 @@ type Dialect interface {
 	IsDeadlock(err error) bool
+
+	// modifier is only useful for sqlite
+	ToTimestamp(d1 string, modifier string) string
```
Given my suggestion of storing and comparing epochs instead of DateTime this can be removed.

marefr reviewed (changes requested):
We should store api key expiration as epochs instead of DateTime in database.

papagian committed:
Frontend fixes

Add checks for empty/undefined/null values

papagian committed:
Change expires column from datetime to integer

papagian committed:
Restrict api key duration input

It should be interval not number

aocenas commented on pkg/services/sqlstore/apikey_test.go:
```diff
@@ -27,5 +29,91 @@ func TestApiKeyDataAccess(t *testing.T) {
+
+		t.Run("Add a key with invalid lifespan", func(t *testing.T) {
+			//expires in one day
+			cmd := m.AddApiKeyCommand{OrgId: 1, Name: "key-with-invalid-lifespan", Key: "asd3", SecondsToLive: -3600}
```
Shouldn't this error out if the value is invalid? If it doesn't right now what will happen? Create a key with expiration in the past?

papagian replied:
No, it creates a key that does not expire. So it actually ignores the invalid value for SecondsToLive.

aocenas replied:
Ah ok right, small nitpick then, maybe we could change the test name to something like "Add key with negative lifespan". I am usually expecting some error if something is invalid.

papagian replied:
maybe it's better to return error in that case

marefr replied:
Don't have all the context here, but one way we've handles this elsewhere is returning a typed error and then if validation error returned we can in the api make sure to return a 400 error instead of 500 error (for unhandled/run time error). I think it's more explicit to return an error here and notify the user.

How is this handled in UI? Can you input a negative timespan? As a general rule UI and API should provide same experience and easiest is that UI respects any error messages returned from the api and adds input validation on top, if needed, to guide the user and provide a better user experience.

papagian replied:
> How is this handled in UI? Can you input a negative timespan?

No, the UI does not allow this.

marefr replied:
Then the API should return an error as well.

aocenas reviewed (approved):
Nice work, looks solid to me.

papagian committed:
AddApiKey must complain if SecondsToLive is negative

torkelo commented:
Great job!

marefr reviewed (changes requested):
See comments.

papagian committed:
Declare ErrInvalidApiKeyExpiration

marefr commented on pkg/api/apikey.go:
```diff
@@ -65,7 +65,7 @@ func (hs *HTTPServer) AddAPIKey(c *m.ReqContext, cmd m.AddApiKeyCommand) Respons
 
 	if err := bus.Dispatch(&cmd); err != nil {
-		if err.Error() == "Negative value for SecondsToLive" {
+		if err == m.ErrInvalidApiKeyExpiration {
```
Should be `models.ErrInvalidApiKeyExpiration` so the import at the top shouldn't be aliased. Think we decided that, but can't find any documentation about it.

marefr commented on pkg/services/sqlstore/apikey.go:
```diff
@@ -44,7 +43,7 @@ func AddApiKey(cmd *m.AddApiKeyCommand) error {
 			expires = &v
 		} else if cmd.SecondsToLive < 0 {
-			return fmt.Errorf("Negative value for SecondsToLive")
+			return m.ErrInvalidApiKeyExpiration
```
Should be `models.ErrInvalidApiKeyExpiration` so the import at the top shouldn't be aliased. Think we decided that, but can't find any documentation about it.

papagian replied:
ok, but the alias is used also in other statements in the same file
should I change all of them and remove the alias completely?

marefr replied:
Yes please remove `m`-alias in all your changed files

marefr commented on pkg/services/sqlstore/apikey_test.go:
```diff
@@ -66,7 +66,7 @@ func TestApiKeyDataAccess(t *testing.T) {
 			cmd := m.AddApiKeyCommand{OrgId: 1, Name: "key-with-negative-lifespan", Key: "asd3", SecondsToLive: -3600}
 			err := AddApiKey(&cmd)
-			assert.EqualError(t, err, "Negative value for SecondsToLive")
+			assert.EqualError(t, err, m.ErrInvalidApiKeyExpiration.Error())
```
Should be `models.ErrInvalidApiKeyExpiration` so the import at the top shouldn't be aliased. Think we decided that, but can't find any documentation about it.

marefr reviewed (changes requested):
Great :+1: Requested some additional code style changes

marefr commented on conf/defaults.ini:
```diff
@@ -505,6 +505,9 @@ global_dashboard = -1
 global_api_key = -1
 
+# global limit of api_key seconds to live before expiration
+global_api_key_max_seconds_to_live = -1
```
It's a bit weird that this is in the `[quota]` section. This should be moved to either `[security]` or `[auth]` included with updated [configuration documenation](https://github.com/grafana/grafana/blob/master/docs/sources/installation/configuration.md).

marefr reviewed (changes requested):
Found some additional things, see comments.

Also, [auth http api documentation](https://github.com/grafana/grafana/blob/master/docs/sources/http_api/auth.md) needs to be updated with updated request/response and parameters.

papagian committed:
Move configuration to auth section

papagian committed:
Update docs

papagian committed:
Eliminate alias for models in modified files

marefr commented on docs/sources/http_api/auth.md:
```diff
@@ -77,12 +77,14 @@ Content-Type: application/json
     "name": "API",
-    "role": "Admin"
+    "role": "Admin",
+    "expiration": null
```
Remove 

```
,
"expiration": null
```

marefr commented on pkg/models/apikey.go:
```diff
@@ -64,7 +68,8 @@ type GetApiKeyByIdQuery struct {
+	Id         int64      `json:"id"`
+	Name       string     `json:"name"`
+	Role       RoleType   `json:"role"`
+	Expiration *time.Time `json:"expiration"`
```
Use the following to omit empty fields being returned as null in the api response:

```
Expiration *time.Time `json:"expiration,omitempty"`
```

papagian replied:
GTK

marefr commented on pkg/services/sqlstore/apikey_test.go:
```diff
@@ -1,31 +1,118 @@
 
 		Convey("Given saved api key", func() {
-			cmd := m.AddApiKeyCommand{OrgId: 1, Name: "hello", Key: "asd"}
+			cmd := models.AddApiKeyCommand{OrgId: 1, Name: "hello", Key: "asd"}
```
Remove usage of convey in this file, see https://github.com/grafana/grafana/blob/master/pkg/README.md#reduce-the-use-of-goconvey

marefr commented on pkg/services/sqlstore/apikey_test.go:
```diff
@@ -1,31 +1,118 @@
+			assert.Nil(t, err)
+
+			// sleep for 1s
+			time.Sleep(time.Second)
```
Since you're using `mockTimeNow` you shouldn't need to use `time.Sleep`. You should be able to to change the "stubbed" time to one second in the future.

marefr commented on pkg/services/sqlstore/migrations/apikey_mig.go:
```diff
@@ -78,4 +78,12 @@ func addApiKeyMigrations(mg *Migrator) {
+	}))
+
+	mg.AddMigration("Add index api_key.expires", NewAddIndexMigration(apiKeyV2, &Index{
+		Cols: []string{"expires"}, Type: IndexType,
```
Not sure if this index is suitable since you're always filtering by `org_id` when filtering by `expires`. So maybe the index should be `org_id` + `expires`?

Or just skip it altogether given that it shouldn't be that many rows in this table.

What do you think?

papagian replied:
maybe I remove the index for now and we add it in the future if we experience any performance issues because of this

marefr replied:
Sounds good :+1:

marefr commented on public/app/features/api-keys/ApiKeysPage.tsx:
```diff
@@ -130,6 +162,17 @@ export class ApiKeysPage extends PureComponent<Props, any> {
+    }
+    date = isDateTime(date) ? date : dateTime(date);
+    format = format || 'YYYY-MM-DD HH:mm:ss';
+    const timezone = store.getState().user.timeZone;
```
There's a selector called `getTimeZone` [here](https://github.com/grafana/grafana/blob/master/public/app/features/profile/state/selectors.ts#L3) that's suggested to be used instead of traversing the state tree in a component.

marefr commented on public/app/features/api-keys/ApiKeysPage.tsx:
```diff
@@ -130,6 +162,17 @@ export class ApiKeysPage extends PureComponent<Props, any> {
 
+  formatDate(date, format?) {
+    if (!date) {
+      return '';
```
Maybe show `No expiration date instead of empty string?`

marefr reviewed (changes requested):
Great work! Now I have tried it out and it works as expected :+1: 

Some minor things left before we can merge this, see comments.

papagian committed:
Omit expiration from api response if empty

papagian committed:
Eliminate Goconvey from test file

papagian committed:
Fix test

Do not sleep, use mocked timeNow() instead

papagian committed:
Remove index for expires from api_key table

The index should be anyway on both org_id and expires fields.
However this commit eliminates completely the index for now
since not many rows are expected to be in this table.

papagian committed:
Use getTimeZone function

papagian committed:
Minor change in api key listing

The frontend should display a message instead of empty string
if the key does not expire.

marefr reviewed (approved):
Great work!
